### PR TITLE
circulation category: negative availability

### PR DESCRIPTION
**Pull request by zannkukai**
* Adds 'disable_circulation' field into the JSON schema. When enable,
  all items related to this circulation category could not be relevant
  for any circulation operations.
* Adds 'displayed_status' field. Define the label to display in case of
  'disable_circulation' is set to True (multi_language field).
* Adds 'circulation_information' field. Define the label to display into
  the public interface (multi_language field).
* Updates indexing behavior if negative availability is updated : we
  need to reindex all related items is asynchronous batch process.

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Review by benerken — APPROVED** [review]

File: rero_ils/modules/item_types/jsonschemas/item_types/item_type-v0.0.1.json [hunk 1]
```diff
@@ -90,6 +94,124 @@
...
+        "propertiesOrder": [
+          "language",
+          "label"
+        ],
+        "properties": {
+          "language": {
+            "$ref": "https://ils.rero.ch/schemas/common/ui-languages-v0.0.1.json#/language"
+          },
+          "label": {
+            "type": "string",
```
Comment (jma) [inline rero_ils/modules/item_types/jsonschemas/item_types/item_type-v0.0.1.json:125]: description?

Comment (jma) [reply rero_ils/modules/item_types/jsonschemas/item_types/item_type-v0.0.1.json:125]: Can you check the order of the properties in the JSONSchema: https://github.com/rero/developer-resources/blob/master/data/jsonschema.md

File: rero_ils/modules/item_types/listener.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            # reindex into a asynchronous celery queue.
+            item_uuids = []
+            search = ItemsSearch()\
+                .filter('bool', should=[
+                    Q('match', item_type__pid=record.pid),
+                    Q('match', temporary_item_type__pid=record.pid)
+                ]) \
+                .source(['pid']).scan()
+            for hit in search:
+                item_uuids.append(Item.get_id_by_pid(hit.pid))
```
Comment (jma) [inline rero_ils/modules/item_types/listener.py:44]: You can get probably the uuid in the elasticsearch `_id` property.

File: rero_ils/modules/items/api/circulation.py [hunk 3]
```diff
@@ -1242,8 +1232,33 @@ def get_loan_states_for_an_item(self):
...
-        return self.item_has_active_loan_or_request() == 0
+        """Get availability for item.
+
+        An item is 'available' if there are no related request/active_loan and
+        if the related circulation category doesn't specify a negative
+        availability.
+        """
+        if self.item_has_active_loan_or_request() > 0:
+            return False
+        if self.circulation_category.get('negative_availability', False):
```
Comment (jma) [inline rero_ils/modules/items/api/circulation.py:1243]: `False` can be removed

File: rero_ils/modules/items/api/circulation.py [hunk 4]
```diff
@@ -1242,8 +1232,33 @@ def get_loan_states_for_an_item(self):
...
+            return False
+        return True
+
+    @property
+    def availability_text(self):
+        """Availability text to display for an item."""
+        circ_category = self.circulation_category
+        if circ_category.get('negative_availability', False):
```
Comment (jma) [inline rero_ils/modules/items/api/circulation.py:1251]: `False` can be removed.

File: rero_ils/modules/items/api/circulation.py [hunk 5]
```diff
@@ -1242,8 +1232,33 @@ def get_loan_states_for_an_item(self):
...
+            messages = circ_category.get('displayed_status', [])
+            messages.append({
+                'language': 'default',
+                'label': circ_category.get('name')
+            })
+            return messages
```
Comment (jma) [inline rero_ils/modules/items/api/circulation.py:1257]: you can have something like
```python
return circ_category.get('displayed_status', []) + [{
                'language': 'default',
                'label': circ_category.get('name')
            }]
```

**Review by jma — CHANGES_REQUESTED** [review]

File: rero_ils/modules/item_types/api.py [hunk 6]
```diff
@@ -75,7 +75,7 @@ def extended_validation(self, **kwargs):
         """
         online_type_pid = self.get_organisation().online_circulation_category()
         if self.get('type') == 'online' and online_type_pid and \
-                self.pid != online_type_pid:
+           self.pid != online_type_pid:
```
Comment (rerowep) [inline rero_ils/modules/item_types/api.py:78]: This is not recommend in pep8:
`E125 continuation line with same indent as next logical line`

**Review by rerowep — APPROVED** [review]

**Review by lauren-d — APPROVED** [review]

File: rero_ils/modules/item_types/listener.py (hunk 2 above)
Comment (zannkukai) [reply rero_ils/modules/item_types/listener.py:44]: you're right : `hit.meta.id` ;-)

**Pull request merged**